Russian literary braille table

Futyn-Maker opened the pull request:
Hi all! 

Now I'm working to improve ru-litbrl.ctb, but I understand that I need help. 

What have I already done? 
- I've recycled ru-litbrl.ctb very a lot. Now it's more correct; 
- I've created the new table - "ru-math.ctb". Russian Braille math system is separate of international system - it's bit same as Marburg, but we have a lot of different things. Now this table contains definetions and rules for Unicode math operators, fractions, superscripts, subscripts and more. 
- I've written small test (22 tests) to ru-litbrl.ctb. It describes Russian literary and math rules. 

What I haven't been able to do? 

Now I have 10 failure tests out of 22. It includes checking indicating of capital letters, direct and dialog speech and math rules. 
I cannot fix it because I don't understand about working context/correct and multipass opcodes. I read documentation, but my experiments with these opcodes were failure. Maybe I didn't understand documentation correct because I'm not very well in English.

I'll glad if somebody will explaine me about making difficult rules or help to fixed it in my table. 

Andrey

Futyn-Maker commented:
Oops. I've created the new branch special for these commits, but the old commits are here... I cannot find out how to delete it :( Sorry

bertfrees commented:
@Futyn-Maker It's OK. Thanks for making the PR.

bertfrees commented:
@Futyn-Maker Nice test. I'll see if I can make some of the tests pass.

Note to @egli: I rebased the branch and pushed it [here](https://github.com/liblouis/liblouis/compare/ru-litbrl) (couldn't push it to Andrey's branch).

Futyn-Maker commented:
Thanks.
The needest things is putting dot 6 before punctuations after fractions, put dots 34 before superscripts and 16 before subscripts. Also the big problems are indicating capital letters of different alphabets and direct speech and dialogs.

bertfrees commented:
@Futyn-Maker I don't know if I will be able to look at it before the next release. I have some other tasks on my list. But you may have some luck asking again on the mailing list if anyone else is prepared to help.

Futyn-Maker commented:
OK, thanks. I'll write to list bit later.

bertfrees commented:
I have rebased this on master: https://github.com/liblouis/liblouis/compare/ru-litbrl

Futyn-Maker commented:
OK, thanks. Now i haven't enough time to continue working on this table, I hope, I'll try it later.

Futyn-Maker commented:
Hello @bertfrees!
I've decided to connect ru-litbrl.ctb and ru-math.ctb. I've removed tests that are don't relevant anymore, moreover, I've achieved some progress with indication of capital letters. Now I have only one problem: the first Cyrillic capital letter after Latin letter isn't indicated like in the test "HaH".

The problem is that it's tthree context-rules which don't work together.

I'll very glad if You help me with it.

Thanks!

Andrey

bertfrees commented:
How come those tests not relevant anymore?

Futyn-Maker commented:
> How come those tests not relevant anymore?

First of all, I've temporarily gave up from creating the Russian math table. It should be done within the LiblouisUDTML and MathPlayer, but it's more difficult.
In addition, I've removing tests of direct speech because of difficulties of this rule in our Braille system.

bertfrees commented:
@Futyn-Maker I have worked on the context rules. All your tests pass now. Check out my branch: https://github.com/liblouis/liblouis/compare/ru-litbrl. Note that I also moved ru-litbrl.yaml into ru.yaml.

The bad news is that a bunch of other tables that are based on ru-litbrl.ctb have failing tests. Maybe you can have a look first? These are the problematic tests. Some are duplicates.

1. Bashkir (ba.yaml)

   - Input:    'Айгөл арып, хәлдән тайып мәктәптән ҡайтып керҙе.'
     Expected: '⠁⠯⠛⠣⠇ ⠁⠗⠮⠏⠂⠓⠜⠇⠙⠜⠝ ⠞⠁⠯⠮⠏ ⠍⠜⠅⠞⠜⠏⠞⠜⠝ ⠩⠁⠯⠞⠮⠏ ⠅⠑⠗⠬⠑⠲'
     Received: '⠘⠁⠯⠛⠣⠇ ⠁⠗⠮⠏⠂⠓⠜⠇⠙⠜⠝ ⠞⠁⠯⠮⠏ ⠍⠜⠅⠞⠜⠏⠞⠜⠝ ⠩⠁⠯⠞⠮⠏ ⠅⠑⠗⠬⠑⠲'

   - Input:    'Этот подъём был круче, тяжелее прежнего.'
     Expected: '⠪⠞⠕⠞ ⠏⠕⠙⠷⠡⠍ ⠃⠮⠇ ⠅⠗⠥⠟⠑⠂⠞⠫⠚⠑⠇⠑⠑ ⠏⠗⠑⠚⠝⠑⠛⠕⠲'
     Received: '⠘⠪⠞⠕⠞ ⠏⠕⠙⠷⠡⠍ ⠃⠮⠇ ⠅⠗⠥⠟⠑⠂⠞⠫⠚⠑⠇⠑⠑ ⠏⠗⠑⠚⠝⠑⠛⠕⠲'

   - Input:    '1, 3, 5, 7, 11, 13, 17, 19, 23'
     Expected: '⠼⠁⠂⠼⠉⠂⠼⠑⠂⠼⠛⠂⠼⠁⠁⠂⠼⠁⠉⠂⠼⠁⠛⠂⠼⠁⠊⠂⠼⠃⠉'
     Received: '⠼⠁⠠⠂⠼⠉⠠⠂⠼⠑⠠⠂⠼⠛⠠⠂⠼⠁⠁⠠⠂⠼⠁⠉⠠⠂⠼⠁⠛⠠⠂⠼⠁⠊⠠⠂⠼⠃⠉'

2. Ukrainian (uk.yaml)

   - Input:    'Вона мріє про зів’яле листя на ґанку, але її мрії не збуваються.'
     Expected: '⠺⠕⠝⠁ ⠍⠗⠽⠜ ⠏⠗⠕ ⠵⠽⠺⠄⠫⠇⠑ ⠇⠊⠎⠞⠫ ⠝⠁ ⠻⠁⠝⠅⠥⠂⠁⠇⠑ ⠹⠹ ⠍⠗⠽⠹ ⠝⠑ ⠵⠃⠥⠺⠁⠳⠞⠾⠎⠫⠲'
     Received: '⠘⠺⠕⠝⠁ ⠍⠗⠽⠜ ⠏⠗⠕ ⠵⠽⠺⠄⠫⠇⠑ ⠇⠊⠎⠞⠫ ⠝⠁ ⠻⠁⠝⠅⠥⠂⠁⠇⠑ ⠹⠹ ⠍⠗⠽⠹ ⠝⠑ ⠵⠃⠥⠺⠁⠳⠞⠾⠎⠫⠲'

   - Input:    'Этот подъём был круче, тяжелее прежнего.'
     Expected: '⠪⠞⠕⠞ ⠏⠕⠙⠷⠡⠍ ⠃⠮⠇ ⠅⠗⠥⠟⠑⠂⠞⠫⠚⠑⠇⠑⠑ ⠏⠗⠑⠚⠝⠑⠛⠕⠲'
     Received: '⠘⠪⠞⠕⠞ ⠏⠕⠙⠷⠡⠍ ⠃⠮⠇ ⠅⠗⠥⠟⠑⠂⠞⠫⠚⠑⠇⠑⠑ ⠏⠗⠑⠚⠝⠑⠛⠕⠲'

   - Input:    '1, 3, 5, 7, 11, 13, 17, 19, 23'
     Expected: '⠼⠁⠂⠼⠉⠂⠼⠑⠂⠼⠛⠂⠼⠁⠁⠂⠼⠁⠉⠂⠼⠁⠛⠂⠼⠁⠊⠂⠼⠃⠉'
     Received: '⠼⠁⠠⠂⠼⠉⠠⠂⠼⠑⠠⠂⠼⠛⠠⠂⠼⠁⠁⠠⠂⠼⠁⠉⠠⠂⠼⠁⠛⠠⠂⠼⠁⠊⠠⠂⠼⠃⠉'

3. Uzbek (uz.yaml)

   - Input:    'Hayirli kech'
     Expected: '⠹⠁⠯⠊⠗⠇⠊⠀⠅⠑⠉⠹'
     Received: '⠨⠹⠁⠯⠊⠗⠇⠊⠀⠅⠑⠉⠹'

   - Input:    'Рождество байрамингиз қутлуғ бўлсин'
     Expected: '⠗⠕⠚⠙⠑⠎⠞⠺⠕⠀⠃⠁⠯⠗⠁⠍⠊⠝⠛⠊⠵⠀⠽⠥⠞⠇⠥⠻⠀⠃⠧⠇⠎⠊⠝'
     Received: '⠘⠗⠕⠚⠙⠑⠎⠞⠺⠕⠀⠃⠁⠯⠗⠁⠍⠊⠝⠛⠊⠵⠀⠽⠥⠞⠇⠥⠻⠀⠃⠧⠇⠎⠊⠝'

4. There is one more, namely spaces.yaml. The failing tests in this file are all about the Russian table though. I have therefore moved them to ru.yaml.

   - Input:    '\x0009'
     Expected: virtual dot 9
     Received: ' ' (space)

   - Input:    '\x00a0'
     Expected: virtual dot a
     Received: ' ' (space)

   - Input:    '\x000c'
     Expected: ' ' (space)
     Received: '⠄⡳⠭⠴⠴⠴⠉⠄'

Futyn-Maker commented:
@bertfrees Great! Many thanks for improve the table, it seems to work as expected now.

Speaking about the test, yes, it are, of course, failur because we've changed the alphabet indicators. I hope I'll find time today to correct these tests. Should I creat new PR with master branch?

And also, about spaces. As I understand, tests you've written about is about ru.ctb (computer braille). They aren't relevant for ru-litbrl.ctb.

Today I should add metadata, some other symbols (arrows and additional letters) and look at test-files of other tables.

Andrey

Futyn-Maker commented:
And one another thing. What about the license? As I know, this table is licensed by LGPL version 3, while other tables are licensed by LGPL 2.1. I understand about the licenses badly, do we have any problems connected with it?

bertfrees commented:
> Speaking about the test, yes, it are, of course, failur because we've changed the alphabet indicators. I hope I'll find time today to correct these tests.

So were all these tests wrong before, and they have been fixed by your change? How should I interpret this? Where the tests wrong because the creator wanted them to pass even though Liblouis behaved wrong, or did the rules actually change?

I mean, the changes seem reasonable. And I fully trust you that you are doing the right thing. I just have a slight difficulty with not knowing whether this is actually based on the rules that are taught to the majority of the Russian population. So I would really like to see this backed by some documentation.

I'm also not quite sure that we can just change the tests for all the other tables that are based on Russian. For example, in Uzbekistan apparently no capital indicators are used. This is mentioned explicitly in the test file.

> Should I creat new PR with master branch?

Best would be if you could base your new changes on my latest version. It's possible to just use the same PR if you force push your new branch. Or you can merge your current Tables/Russian branch with my branch, and then do a simple push. If all this is too difficult you may also create a new PR.

> And also, about spaces. As I understand, tests you've written about is about ru.ctb (computer braille). They aren't relevant for ru-litbrl.ctb.

No, they were written for ru-litbrl.ctb (or ru.tbl, which is basically the same). The space tests for ru.ctb had already been moved to ru.yaml before.

Futyn-Maker commented:
> So were all these tests wrong before, and they have been fixed by your change? How should I interpret this? Where the tests wrong because the creator wanted them to pass even though Liblouis behaved wrong, or did the rules actually change?
> I mean, the changes seem reasonable. And I fully trust you that you are doing the right thing. I just have a slight difficulty with not knowing whether this is actually based on the rules that are taught to the majority of the Russian population. So I would really like to see this backed by some documentation.
> I'm also not quite sure that we can just change the tests for all the other tables that are based on Russian. For example, in Uzbekistan apparently no capital indicators are used. This is mentioned explicitly in the test file.

To be honest, it's really difficult story.

First of all, I should say that Russian Braille is documented weakly. In fact, we have only (https://github.com/liblouis/braille-specs/blob/master/russian/%D0%A0%D1%83%D0%BA%D0%BE%D0%B2%D0%BE%D0%B4%D1%81%D1%82%D0%B2%D0%BE%20%D0%BF%D0%BE%20%D0%B2%D1%8B%D0%BF%D1%83%D1%81%D0%BA%D1%83%20%D0%B1%D1%80%D0%B0%D0%B9%D0%BB%D0%B5%D0%B2%D1%81%D0%BA%D0%B8%D1%85%20%D0%B8%D0%B7%D0%B4%D0%B0%D0%BD%D0%B8%D0%B9%20%D0%BC%D0%B0%D1%81%D1%81%D0%BE%D0%B2%D0%BE%D0%B3%D0%BE%20%D1%80%D0%B0%D1%81%D0%BF%D1%80%D0%BE%D1%81%D1%82%D1%80%D0%B0%D0%BD%D0%B5%D0%BD%D0%B8%D1%8F.pdf)[Mass-braille distribution guide] which is about imbossing paper-books.
This guide say that we shouldn't indicate capital letters in most cases, where it's obvious. According to it, capital mode is need to figure out Ambiguities.

But the problem is that computer can't find the places where it's neccesary to put the capital indicator. So I think it's better to indicate capital letters anywhere than we'd totally ignore them. Besides, the same system is using on IOS by years.

Speaking about other languages based on the Russian Braille, i don't think they have any different rules. The problem is that there's almost no any documents. The attitute to capital letters were formed many years ago, but now, when we have computers and Braille displays, it's no so Unambiguously.
I know quite many people who always indicates capital letters when writing, even on paper, but now I can't sure that they consist of a majority.

Andrey

bertfrees commented:
Thanks for the information, that clarifies things for me. I find it quite astonishing that for such a huge country like Russia there is so limited documentation. But OK, it's the way it is...

It would be helpful if you could include this explanation (the reasons for wanting to indicate capitals or not indicate them) in the YAML files. Preferably I would like to keep both the tests with and without capital indication. The tests that don't pass can then be commented out (or marked as failing).

From what I understand, maybe there is even the need for two versions of the Russian table? It sounds like there are people who would rather not have capitals indicated, especially when we're talking about braille on paper.

We also make this distinction in Danish, and recently also for German.

What do you think?

Futyn-Maker commented:
> From what I understand, maybe there is even the need for two versions of the Russian table? It sounds like there are people who would rather not have capitals indicated, especially when we're talking about braille on paper.
> 
> We also make this distinction in Danish, and recently also for German.
> 
> What do you think?

Yes, It makes sense, although nobody uses Liblouis-based programs for preparation texts for imbosing in Russia - all libraries work with the DBT. The majority of the table's users is the NVDA and JAWS users with Braille displays.

Now I run the survey in Russian communities and ask people if they prefer read texts with capitalizing or not. Maybe, if opinions differ a lot, I'll make the other version of this table.

bertfrees commented:
> The majority of the table's users is the NVDA and JAWS users with Braille displays.

But still, since there are so many Russians, I bet there are also a lot of people who want to stick to what they're used to, even when using a computer.

Also we don't just need to accept the Duxbury domination. Liblouis is used for production of braille on paper in many countries so why not Russia?

> Now I run the survey in Russian communities

Great idea. Thanks!

Futyn-Maker commented:
Well, I've got about 7 opinions that capitalizing is necessary and only one that uppercase letters shouldn't be indicated. I suggest including the version with indicating upercases in this release and than think about creating an altirnative table.

I've also written to Igor Poretsky, let's hope he'll give an answer. :)

I'll pushing the table here bit later. I want to add Greek mode and some symbols.

Futyn-Maker commented:
Well, I've found the new problem (yes, another one!) and created two new tests which now is failure.

@bertfrees Can You explaine why the rule `always ,\s 2` doesn't work after capital letters? I find it quite strange, it is the "always"-rule.

About other changes. I've added the Greek mode to indicate Greek letters, too. I've also bit changed the rule and the test of capitalizing cyrillics, because in fact, the word which consists of several capital Cyrillic letters should be marked with 45-45. I've also added tests without capitalizing.

bertfrees commented:
> Can You explaine why the rule always ,\s 2 doesn't work after capital letters?

Hmm, it's because of the "Exit uppercase mode" rules that I added. As a matter of fact, this was messing up *all* the `always` rules and other translation rules. I have fixed it now.

Futyn-Maker commented:
Great! Many thanks, while I don't understand all aspects of Liblouis's working.

Should I add in other languags' test-files the same tests, without capitalizing?

Futyn-Maker commented:
@bertfrees You've removed the swapcd updigit class, is it a mistake?

bertfrees commented:
> Should I add in other languages' test-files the same tests, without capitalizing?

Yes, would be great.

> You've removed the swapcd updigit class, is it a mistake?

No, I removed it because the class is no longer used anywhere.

Futyn-Maker commented:
> No, I removed it because the attribute is no longer used anywhere.

```noback context $d %updigit#2=0```
This rule are still in the table (I see in your branch). Can I just remove?

bertfrees commented:
Oops, that was a mistake. I meant to remove it. But now I notice it doesn't work without this rule. Don't understand what is happening. OK, you can put back the updigit class. But it shouldn't be in the "Exit uppercase mode" section.

Futyn-Maker commented:
I've pushed Your changes here, but with putting back updigit-class.
It doesn't seem to have any problems now...

Futyn-Maker commented:
I also have some problems with others tests due to virtual dot 9. We don't use it in ru-litbrl.ctb anymore, MB remove it from other tables based on it, too? I have failurs tests due to this...

poretsky commented:
Of course, I have no objections concerning licensing and re-licensing
of this stuff. But I am absolutely disagree that these tables are
generally not used for braille embossing preparation. I started this
work just to get a versatile braille printing technology and up to the
moment use it exclusively for this purpose. Moreover, as far as I
know, Everest braille embossers manufactured by Index Braille include
Liblouis software along with its tables as embedded braille
preparation means.

Futyn-Maker commented:
@poretsky Hello and thank You for your answer!

When I was talking about this problem, I meaned libraries and other printing-houses. Of course, I know that, for example, in Linux-systems Liblouis-based programs is the only way to prepare texts for embossing. But now Your table is included in a lot of others, that people use with Braille displays and I think it's necessary to do these changes due to this fact. Your comment proves that we need two different tables.
Excepted indication of capital letters and adding new symbols, this table have other changes. For example, the dot 6 after digits and before punctuations.

I also don't know how often the Liblouis-based programs for working with texts are updated and if it is possible to change default tables easy.

Futyn-Maker commented:
Basically, we have another table which now is used by NVDA - ru-ru-g1.utb. This table was created by Viewplus for TSS, but contains a lot of bugs.
I could copy recycled ru-litbrl.ctb to this table, but I think we'll have problems with copyright in this case, right?

Initially I was working with ru-litbrl.ctb because this table had already contained a lot of needful rules.

Futyn-Maker commented:
I'm looking at the Uzbek test-file... It seems they really not use capitalizing... It's really hard to know more when there're no any documents.

bertfrees commented:
OK I think the best way forward is to create two versions of the Russian table. Then we can base Uzbek, etc. on the version without capital indication for now. (This can be changed later, but we need more time to discuss it with the maintainers of those tables.)

The version with capital indication should include the other one and add some rules to it. From what I have seen from the table this should be doable and not much work. Copying ru-litbrl.ctb and modifying the copy would also be possible, there are no copyright issues with this, however from the perspective of maintainability it is a bad idea.

Regarding your concern about the time it could take for your improvements to end up in products: The manufacturers need to be encouraged to update Liblouis frequently, and to pick the tables that suit their product best in case there are multiple variants of a table. This is not an issue specific to Russian. If you fear that braille displays that currently use the default ru-litbrl.ctb might not switch to the new variant, you should make it very clear in the release notes that the new table is more suited for braille displays than the current one.

bertfrees commented:
By the way the metadata for the table with capital indicators should be `#+variant: detailed`. We have discussed the same issue in the context of other languages and this is what came out of that discussion.

Futyn-Maker commented:
> Regarding your concern about the time it could take for your improvements to end up in products: The manufacturers need to be encouraged to update Liblouis frequently, and to pick the tables that suit their product best in case there are multiple variants of a table. This is not an issue specific to Russian. If you fear that braille displays that currently use the default ru-litbrl.ctb might not switch to the new variant, you should make it very clear in the release notes that the new table is more suited for braille displays than the current one.

Anyway, I'm going to open an issue at least in NVDA to replace ru-ru-g1.utb with ru-litbrl.ctb in the GUI. We'll have to make something with ru-ru-g1.utb than - either remove or modify it... The problem that a lot of programs is Abandonware now.

I'm also not sure what table should be main - with capitalizing or not. I mean what table should include in the second table. I think the table with indication of capitals rather be more popular than without it.

Futyn-Maker commented:
It's the correction of tests and tables. I've remove the virtual dot 9 and corrected tests. While I don't know what to do with the Uzbek table and test-file.

bertfrees commented:
> I'm also not sure what table should be main - with capitalizing or not. [...] I think the table with indication of capitals rather be more popular than without it.

It's not really a question about main or secondary. They are just two variants.

> I mean what table should include in the second table.

This is another question. This is about what is most practical. It's probably more practical to include the table without capitalization in the table with capitalization.

>  I've remove the virtual dot 9.

I hadn't even noticed you had removed the dot 9 in ru-litbrl.ctb. Why did you do that? I thought it was a useful feature.

> While I don't know what to do with the Uzbek table and test-file.

As I said before I would not change the behavior of the Uzbek, Ukraïnian (and maybe even Bashkir) for now. This can be sorted out later.

Futyn-Maker commented:
> I hadn't even noticed you had removed the dot 9 in ru-litbrl.ctb. Why did you do that? I thought it was a useful feature.
And what for is it needed for table with forward flag? I think it's an just excess.

> As I said before I would not change the behavior of the Uzbek, Ukraïnian (and maybe even Bashkir) for now. This can be sorted out later.
The addition difficulty with the Uzbek Braille is the fact that they use latin alphabet, too. As for the Ukrainian Braille, it's really as close as possible to the Russian Braille and problems with capitalizing is the same. I can speak nothing about the Bashkir one.

bertfrees commented:
> And what for is it needed for table with forward flag? I think it's an just excess.

I don't think it was just excess, it was nice to have. Admittedly it slightly complicated the table (but really only slightly), but it gave you additional information on what the braille dots represented. It allowed you to use the ru-ru.dis display table to render Cyrillic and Latin differently.

Futyn-Maker commented:
> I don't think it was just excess, it was nice to have. Admittedly it slightly complicated the table (but really only slightly), but it gave you additional information on what the braille dots represented. It allowed you to use the ru-ru.dis display table to render Cyrillic and Latin differently.

TBH, I've never understood the philosophy of display-codes and where it is used... I've remove the dot 9 with removing dot 7 from capital letters (that allowed us to remove the excess context-rule), but if You think the virtual dot 9 is important, I'll put it back. I just saw it in few tables, if at all.

bertfrees commented:
> It allowed you to use the ru-ru.dis display table to render Cyrillic and Latin differently.

I meant ru-letters.dis.

I'm gonna try to make this PR ready for today's release, but I'm not sure I'm gonna make it. There's really not much time left.

bertfrees commented:
@Futyn-Maker I updated my branch. Please have a final look before we merge and release it.

The only thing that is missing now is an entry in the NEWS file. Could you help me? You didn't explain in very much detail what you improved. Can you summarize?

Futyn-Maker commented:
@bertfrees As for me, it's fantastic!  I thought we'd needed much more time to do two versions of the table, but You've really done it fast and well, many thanks for Your help :)

As for news, I think we can be limited with common phrases because changes is very large.

> The table has been recycled a lot according to actual rules of the Russian Braille. A lot of new symbols and rules were added and many bugs were fixed, now it is more convenient for Braille displays' users.

If You aren't against, I'll refrain from commiting something here.

As I see, You've removed ru-ru-g1.utb, I'm absolutely agree with this solution.

bertfrees commented:
Hmm, actually I was thinking about something slightly less general. That is also the problem with your commits. I wasn't really able to distill from your commit messages what actually changed.

We don't need to be specific about which symbols were added. But it would be nice to have an idea about which rules were missing.

It should also be very explicitly mentioned that there is a *new* table with indication of capitals, which we recommend for braille display users. It should be clear that these particular changes are not *automatically* included upon updating.

Futyn-Maker commented:
I hope I forget nothing here.

bertfrees commented:
Thanks!

Futyn-Maker commented:
And yet I've forgotten about this including, while it's quite important.

Futyn-Maker commented:
Well, it have finally been done! My congratulations on the release! And thank You very much for help, I've never done these changes without it. :)

I close this PR due to conflicts. If it's necessary to improve the table, I'll open new one.

Thanks!

Andrey

bertfrees commented:
Thank you for your patience!

bertfrees commented:
Merged in 64f68aa06a244e6836a2c996c04cbd6496f8acd8.

Futyn-Maker commented:
I should say You the same :)
